File: M3Tools/Controls/People/AddressField.vb
Public Class AddressField
	Public Property Address As Types.Address
		Get
			Return Types.Address.Parse(Street, City, State, ZipCode)
		End Get
		Set(value As Types.Address)
			Street = If(value?.Street, "")
			City = If(value?.City, "")
			State = If(value?.State, "")
			ZipCode = If(value?.ZipCode, "")
		End Set
	End Property

	Public Property Street As String
		Get
			If String.IsNullOrWhiteSpace(if_Address1.Text) Then
				' No values present
				Return ""
			End If

			If String.IsNullOrWhiteSpace(if_Address2.Text) Then
				' Only Address1 is filled
				Return if_Address1.Text
			End If

			' Both fields present
			Return String.Join(",", if_Address1.Text, if_Address2.Text)
		End Get
		Set(value As String)
			If value.Contains(",") Then
				Dim parts = value.Split(","c)
				if_Address1.Text = parts(0)
				if_Address2.Text = parts(1)
				Return
			End If

			if_Address1.Text = value
		End Set
	End Property

	Public Property City As String
		Get
			Return if_City.Text
		End Get
		Set(value As String)
			if_City.Text = value
		End Set
	End Property

	Property State As String
		Get
			Return ssf_State.StateCode
		End Get
		Set(value As String)
			If value?.Length >= 2 Then
				value = value.Substring(0, 2)
			End If

			ssf_State.StateCode = value
		End Set
	End Property

	Public Property ZipCode As String
		Get
			Return if_ZipCode.Text
		End Get
		Set(value As String)
			If value?.Length > 5 Then
				value = value.Substring(0, 5)
			End If

			if_ZipCode.Text = value
		End Set
	End Property

	Public ReadOnly Property IsValidAddress As Boolean
		Get
			Return ValidStreet() And ValidCity() And ValidState() And ValidZip()
			'Return (Street <> "" AndAlso City <> "" AndAlso State <> "" AndAlso ZipCode <> "") OrElse (String.IsNullOrEmpty(Street) AndAlso String.IsNullOrEmpty(City) AndAlso String.IsNullOrEmpty(State) AndAlso String.IsNullOrEmpty(ZipCode))
		End Get
	End Property

	Private Function ValidateValue(value As String) As Boolean
		If Not AllEmpty() AndAlso String.IsNullOrWhiteSpace(value) Then
			Return False
		End If

		Return True
	End Function

	Private Function ValidStreet() As Boolean
		If Not ValidateValue(Street) Then
			ep_InvalidAddress.SetError(if_Address1, "A valid street address is required, or you can enter an empty address")
			Return False
		End If

		ep_InvalidAddress.SetError(if_Address1, String.Empty)
		Return True
	End Function
	Private Function ValidCity() As Boolean
		If Not ValidateValue(City) Then
			ep_InvalidAddress.SetError(if_City, "A valid city is required, or you can enter an empty address")
			Return False
		End If

		ep_InvalidAddress.SetError(if_City, String.Empty)
		Return True
	End Function
	Private Function ValidState() As Boolean
		If Not ValidateValue(State) Then
			ep_InvalidAddress.SetError(ssf_State, "A valid state code is required, or you can enter an empty address")
			Return False
		End If

		ep_InvalidAddress.SetError(ssf_State, String.Empty)
		Return True
	End Function
	Private Function ValidZip() As Boolean
		If Not ValidateValue(ZipCode) Then
			ep_InvalidAddress.SetError(if_ZipCode, "A valid zip code is required, or you can enter an empty address")
			Return False
		End If

		ep_InvalidAddress.SetError(if_ZipCode, String.Empty)
		Return True
	End Function

	Private Function AllEmpty() As Boolean
		Return Street <> "" AndAlso City <> "" AndAlso State <> "" AndAlso ZipCode <> ""
	End Function
End Class

File: M3Tools/Controls/People/AddressField.cs
// MAYBE: Use the Address.Display function here instead?
namespace SPPBC.M3Tools
{
	/// <summary>
	/// <inheritdoc/>
	/// </summary>
    public partial class AddressField
    {
		/// <summary>
		/// The address being displayed
		/// </summary>
        public Types.Address Address
        {
            get
            {
                return Types.Address.Parse(Street, City, State, ZipCode);
            }
            set
            {
                Street = (value?.Street) ?? "";
                City = (value?.City) ?? "";
                State = (value?.State) ?? "";
                ZipCode = (value?.ZipCode) ?? "";
            }
        }

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
        public string Street
        {
            get
            {
                if (string.IsNullOrWhiteSpace(if_Address1.Text))
                {
                    // No values present
                    return "";
                }

                if (string.IsNullOrWhiteSpace(if_Address2.Text))
                {
                    // Only Address1 is filled
                    return if_Address1.Text;
                }

                // Both fields present
                return string.Join(",", if_Address1.Text, if_Address2.Text);
            }
            set
            {
                if (value.Contains(","))
                {
                    string[] parts = value.Split(',');
                    if_Address1.Text = parts[0];
                    if_Address2.Text = parts[1];
                    return;
                }

                if_Address1.Text = value;
            }
        }

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
        public string City
        {
            get
            {
                return if_City.Text;
            }
            set
            {
                if_City.Text = value;
            }
        }

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
        public string State
        {
            get
            {
                return ssf_State.StateCode;
            }
            set
            {
                if (string.IsNullOrWhiteSpace(value))
                {
                    return;
                }

                if ((value?.Length) is { } arg1 && arg1 > 2)
                {
                    value = Utils.StateCodeToState(value);
                }

                ssf_State.StateCode = value;
            }
        }

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
        public string ZipCode
        {
            get
            {
                return if_ZipCode.Text;
            }
            set
            {
                if ((value?.Length) is { } arg2 && arg2 > 5)
                {
                    value = value.Substring(0, 5);
                }

                if_ZipCode.Text = value;
            }
        }

		/// <summary>
		/// Whether the address is valid
		/// </summary>
        public bool IsValidAddress
        {
            get
            {
                return ValidStreet & ValidCity & ValidState & ValidZip;
                // Return (Street <> "" AndAlso City <> "" AndAlso State <> "" AndAlso ZipCode <> "") OrElse (String.IsNullOrEmpty(Street) AndAlso String.IsNullOrEmpty(City) AndAlso String.IsNullOrEmpty(State) AndAlso String.IsNullOrEmpty(ZipCode))
            }
        }

		/// <summary>
		/// <inheritdoc/>
		/// </summary>
        public AddressField()
        {
            InitializeComponent();
        }

        private bool ValidateValue(string text)
        {
            if (!AllEmpty && string.IsNullOrWhiteSpace(text))
            {
                return false;
            }

            return true;
        }

        private bool ValidStreet
        {
            get
            {
                if (!ValidateValue(Street))
                {
                    ep_InvalidAddress.SetError(if_Address1, "A valid street address is required, or you can enter an empty address");
                    return false;
                }

                ep_InvalidAddress.SetError(if_Address1, string.Empty);
                return true;
            }
        }

        private bool ValidCity
        {
            get
            {
                if (!ValidateValue(City))
                {
                    ep_InvalidAddress.SetError(if_City, "A valid city is required, or you can enter an empty address");
                    return false;
                }

                ep_InvalidAddress.SetError(if_City, string.Empty);
                return true;
            }
        }

        private bool ValidState
        {
            get
            {
                if (!ValidateValue(State))
                {
                    ep_InvalidAddress.SetError(ssf_State, "A valid state code is required, or you can enter an empty address");
                    return false;
                }

                ep_InvalidAddress.SetError(ssf_State, string.Empty);
                return true;
            }
        }

        private bool ValidZip
        {
            get
            {
                if (!ValidateValue(ZipCode))
                {
                    ep_InvalidAddress.SetError(if_ZipCode, "A valid zip code is required, or you can enter an empty address");
                    return false;
                }

                ep_InvalidAddress.SetError(if_ZipCode, string.Empty);
                return true;
            }
        }

        private bool AllEmpty
        {
            get
            {
                return !(!string.IsNullOrEmpty(Street) && !string.IsNullOrEmpty(City) && !string.IsNullOrEmpty(State) && !string.IsNullOrEmpty(ZipCode));
            }
        }
	}
}
